Add Values and Representation chapter (#1664)
rust-lang/reference: src/types/enum.md

r[type.enum.constructor]
New instances of an `enum` can be constructed with a [struct expression].

r[type.enum.name]
Enum types cannot be denoted *structurally* as types, but must be denoted by
named reference to an [`enum` item].

## Enum values and representation

r[type.enum.value]

r[type.enum.value.intro]
An enum value corresponds to exactly one variant of the enum, and consists of the fields of that variant

> [!NOTE]
> An enum with no variants therefore has no values.

r[type.enum.value.value-padding]
A byte is a [padding][type.union.value.padding] byte of an enum if that byte is not part of the representation of the discriminant of the enum, and in each variant it either:
* Does not overlap with a field of the variant, or
* Overlaps with a padding byte in a field of that variant.

r[type.enum.value.repr]

> The representation of a value of an enum type includes the representation of each field of the variant at the appropriate offsets. When encoding a value of an enum type, each byte which is not use d to store a field of the variant or the discriminant is . In the case of a [`repr(C)`][layout.repr.c.adt] or a [primitive-repr][layout.repr.primitive.adt] enum, the discriminant of the variant is represented as though by the appropriate integer type stored at offset 0.

Reviewer: ```suggestion
The representation of a value of an enum type includes the representation of each field of the variant at the appropriate offsets. When encoding a value of an enum type, each byte that is not used to store a field of the variant or the discriminant is unspecified. In the case of a [`repr(C)`][layout.repr.c.adt] or a [primitive-repr][layout.repr.primitive.adt] enum, the discriminant of the variant is represented as the appropriate integer type stored at offset 0.
```

r[type.enum.constructor]
New instances of an `enum` can be constructed with a [struct expression].

r[type.enum.name]
Enum types cannot be denoted *structurally* as types, but must be denoted by
named reference to an [`enum` item].

## Enum values and representation

r[type.enum.value]

r[type.enum.value.intro]

> An enum value corresponds to exactly one variant of the enum, and consists of the fields of that variant

Reviewer 2: ```suggestion
An enum value corresponds to exactly one variant of the enum, and consists of the fields of that variant.
```

> [!NOTE]
> An enum with no variants therefore has no values.

r[type.enum.value.variant-padding]
A byte is a padding byte in a variant `V` if the byte is not used for computing the discriminant, and the byte would be a padding byte in a struct consisting of the fields of the variant at the same offsets.

r[type.enum.value.value-padding]
A byte is a padding byte of an enum if it is a padding byte in each variant of the enum. A byte that is not a padding byte of an enum is a value byte.

r[type.enum.value.repr]

> The representation of a value of an enum type includes the representation of each field of the variant at the appropriate offsets. When encoding a value of an enum type, each byte which is a padding byte in the variant is set to uninit. In the case of a [`repr(C)`][layout.repr.c.adt] or a [primitive-repr][layout.repr.primitive.adt] enum, the discriminant of the variant is represented as though by the appropriate integer type stored at offset 0.

Reviewer 2: This could be defined without having to define padding bytes, by saying something like:

When encoding an enum value, first the fields of the active variant are encoded at the respective offsets. Bytes which are not contained in any field of the active variant are left uninitialized. Then, the encoding of the discriminant of the active variant is applied to the byte sequence.

Author: It also needs to define padding of the enum as a whole for the purposes of unions, though.

Reviewer 2: Yeah, that is true. I'm not a big fan of the entire concept of "padding" being so pervasive, IMO it is unnecessary... but I guess it is common and many people coming from C/C++ can use prior knowledge to understand it.

> [!NOTE]

> > Most `repr(Rust)` enums will also store a discriminant in the representation of the enum, but the exact placement or type of the discriminant is unspecified, as is the value that represents each variant.

Reviewer 2: ```suggestion
> Most `repr(Rust)` enums will also store a discriminant in the representation of the enum, but the exact placement or type of the discriminant is unspecified, as is the byte sequence that represents each variant.
```
